### PR TITLE
Added a note to sortperm! docstring that setting initialized=true can be dangerous!

### LINKED ISSUE #29179: sortperm! with bad initialization of Int32 indices segfaults
```
julia> r=rand(1000); p=rand(Int64,1000);
julia> sortperm!(p, r; initialized=true)
ERROR: BoundsError: attempt to access 1000-element Array{Float64,1} at index [1165611285912797586]

julia> r=rand(1000); p=rand(Int32,1000);
julia> sortperm!(p, r; initialized=true)

signal (11): Segmentation fault
```
I have no clue why `Int64` catches the bounds-error and `Int32` segfaults.

Also, the documentation of `sortperm!` could be clearer: I managed to misinterpret the `initialized=true` option as "apply the sorting permutation to the preallocated vector" (no idea what I was thinking). Checked on 1.0.0 and 1.1.0-DEV.254, 64bit.

### PR DESCRIPTION
**Pull request by jpsamaroo**
Fixes #29179 

Please close with prejudice if this is not actually desired!

EDIT: @chethega does this addition seem like it would be sufficient, or is there additional information that should be added/changed?

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Especially given that validating the input is probably O(n) while the sortperm op is O(n log(n)). At least for checking that everything is in-bounds. Checking that each index occurs only once is harder but seems strictly easier than sorting all of them. Using a bit vector of the length of the array is possible but seems a bit excessive; I wonder if there's something cleverer to do.

File: base/sort.jl [hunk 1]
```diff
@@ -866,7 +866,8 @@ end
     sortperm!(ix, v; alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward, initialized::Bool=false)
 
 Like [`sortperm`](@ref), but accepts a preallocated index vector `ix`.  If `initialized` is `false`
-(the default), `ix` is initialized to contain the values `1:length(v)`.
+(the default), `ix` is initialized to contain the values `1:length(v)`. Note that if `initialized` is `true`,
+and `ix` is not properly initialized, julia is likely to crash.
```
Comment (chethega) [inline base/sort.jl]: >(the default), `ix` is initialized to contain the values `collect(eachindex(v))`, typically `1:length(v)`. If `initialized` is true, then `ix` is assumed to be an arbitrary permutation of `collect(eachindex(v))`. If this assumption is violated by `ix` not being a permutation, then `sortperm` will return wrong results. If `ix` contains out-of-bounds indices then julia may segfault, if you are lucky; or encode available secrets like passphrases in the result, if you are unlucky.

Ok, just joking. But the info I was missing was that `ix` wants to be initialized to an arbitrary permutation.

>If this assumption is violated by `ix` not being a permutation, then `sortperm` will return wrong results; if it contains oob indices then sortperm will read oob memory.

**Comment by chethega** [comment]
>I wonder if there's something cleverer to do.

First check inbounds (run over input once). Then sort. Then check that no subsequent indices are equal (run over output once).

**Comment by StefanKarpinski** [comment]
The "check that no subsequent indices are equal" part happens after the sort, so if that assumption being violated can lead to crashes in the sorting, then it's too late to check it afterwards. However, I suspect that the in bounds assumption is the only one whose violation can cause a segfault.

**Comment by chethega** [comment]
>However, I suspect that the in bounds assumption is the only one whose violation can cause a segfault.

I think so as well. 

Honestly, my mistake was that I assumed that sortperm worked by pulling a copy of the data, and sorting the copy inplace, applying every exchange to both `ix` and the copy. This resulting semantics would have been "apply the sorting permutation of the input to ix, whatever it is", and the sort would have better cache locality. Jumping to that conclusion from the docstring without looking into the source was not clever of me.

This might still be a better way of computing the sortperm, but would require quite a bit of refactoring of the internal sort functions (because they would need to support applying the permutation to multiple arrays).

**Comment by JeffBezanson** [comment]
Closing since we settled on bounds-checking the index vector.

**Pull request closed without merging**